app/jobs/update_dqt_trn_request_job.rb

```diff
@@ -10,25 +10,51 @@ class StillPending < StandardError
   def perform(dqt_trn_request)
     return if dqt_trn_request.complete?
 
-    response =
-      if dqt_trn_request.initial?
-        DQT::Client::CreateTRNRequest.call(
-          request_id: dqt_trn_request.request_id,
-          application_form: dqt_trn_request.application_form,
-        )
-      else
-        DQT::Client::ReadTRNRequest.call(request_id: dqt_trn_request.request_id)
-      end
+    response = fetch_response(dqt_trn_request)
 
     dqt_trn_request.pending! if dqt_trn_request.initial?
 
     if (trn = response[:trn]).present?
-      ActiveRecord::Base.transaction do
-        dqt_trn_request.application_form.teacher.update!(trn:)
-        dqt_trn_request.complete!
-      end
+      update_teacher_trn(dqt_trn_request, trn)
+      send_award_email(dqt_trn_request)
+      update_application_form(dqt_trn_request)
```

Comment:
I think we should extract this out and think about the failure conditions a bit. I think as it is if the application form update fails we could end up sending multiple emails to the applicant. It think we'd want to do the updates in a transaction and only email if the transaction commits. We could probably do that in here but I think we're going beyond the job's responsibility at that point.

Reply:
My initial implementation had everything in a transaction, but then I was worried about what would happen if creating the email job failed, but the transaction had succeeded (let's say for example the Sidekiq process gets restarted at that exact point). I figured it was better that there's a chance the teachers will receive a duplicate email than not receive the email at all.

I reordered things so the email is sent before the status is updated, but there's a guard clause before sending the email to check we haven't already tried to send one before.

I also wondered about extracting this logic out in to a service, but because of the careful ordering, it felt like the service was then just doing everything the job did.

Happy to talk about it further though!